### DISCUSSION TITLE
Why We Need get Force from Edge feature？

### DISCUSSION THREAD

**Discussion by SyntaxSmith** [MDIL-SNU/SevenNet#66] [Q&A]
In other MLIP models ， we can get a energy for the whole structure firstly, and then we get the force as negetive grad about the energy and atoms' positions, which is the same as physical principle. But I notice this model get force from negetive grad about the energy and edge features between atoms. I want to know why we need change the method to get force? I know Allegro uses the same method, is this method significant for parallel?

**Comment by YutackPark** [comment]
Thanks for your interest. You're correct that in our parallel LAMMPS code, sevennet compute force from edges. 

The reason is that it is much easier to compute and formulate from edge forces, if it is parallel version (using spatial decomposition). 

You may check our JCTC paper. Edge force is involved into the parallel algorithm. (It is a matter of convenience: formulating, and implementing via edge force is much simpler)

Moreover, in PBC, to correctly compute force using autograd, you need to compute relative position vectors considering cell shift (integer times lattice vector) caused by the PBC. It is fine if it is a serial mode. But it becomes complicated if the simulation cell is decomposed for parallelization. In this scenario, computing edge force and postprocessing them is much easier.

Let me know if there is another question or the answer is obscure. I think it is really a good question!

**Reply by SyntaxSmith** [reply]
In another question, I saw that you mentioned if we already have a trained model and want to run it in parallel under LAMMPS, we just need to add the -p parameter during deployment. But I noticed that in the build_equivariant_model function in the trainer, the serial mode model is called. Also, in the ForceOutput class in force_output.py, the method of calculating force is to compute the gradient of energy with respect to the pos_tensor. I want to know that since deploy_parallel does not seem to modify the force output module, where is the operation of calculating the force from the edge implemented?

**Reply by SyntaxSmith** [reply]
In other words, how is the force calculated when we train the model? If the method for calculating the force changes when deploying it in parallel mode, can we still use the trained model as it is?

**Reply by YutackPark** [reply]
That part may confusing as we indeed use different method to compute force. However, as both method (computing from edge force and using autograd to differentiate by positions) give exactly same outputs and there are no model parameters involved, they are interchangeable.

For serial model inference and training, model computes force using autograd (differentiate by positions). For parallel model inference with LAMMPS, model computes force from its edges.

This... confusing scheme arises as we didn't know how to compute stress when force is computed from edges. However, as we figured out the way to compute stress from edge forces, it is subject to change, for consistency. Future version of SevenNet will always compute force and stress from edge forces.

>  want to know that since deploy_parallel does not seem to modify the force output module, 

Here's the relevant code: https://github.com/MDIL-SNU/SevenNet/blob/9f92aaa71e1285dac08a18df5468598034c868da/sevenn/model_build.py#L384
When model is deployed for parallel, `is_parallel=True` is passed to `build_equivariant_model` method. In this case, we don't use force output module written in python. The autograd call for edge force computation is done in `pair_e3gnn/pair_e3gnn_parallel.cpp`.

Even though these complications, I want to emphasize that we have verified both serial and parallel model gives exactly same force outputs. You can verify it yourself if you want.
